src/mds/MDLog.cc

```diff
@@ -651,13 +656,32 @@ void MDLog::trim(int m)
     ceph_assert(segments.size() >= pre_segments_size);
     max_expiring_segments = std::max<unsigned>(max_expiring_segments,segments.size() - pre_segments_size);
   }
-  
+
   map<uint64_t,LogSegment*>::iterator p = segments.begin();
+
+  auto trim_start = ceph::coarse_mono_clock::now();
+  std::optional<ceph::coarse_mono_time> trim_end;
+
+  auto log_trim_counter_start = log_trim_counter.get();
+  auto log_trim_threahold = g_conf().get_val<Option::size_t>("mds_log_trim_threshold");
+
   while (p != segments.end()) {
-    if (stop < ceph_clock_now())
-      break;
+    // throttle - break out of trimmming if we've hit the threahold
+    if (log_trim_counter_start + new_expiring_segments >= log_trim_threahold) {
```

Comment:
Well, I have deliberately not added checks that exists out of this loop yet. This is to discuss [#4](https://tracker.ceph.com/issues/61908#note-4).

Also missing is the heartbeat reset which again is deliberately not added yet, therefore the RFC tag in the PR description.

Reviews and suggestion welcome.

Reply:
You could just do

`trim_counter.get() >= log_trim_threshold`

but I guess your concern may be that if expiring each segment, for some reason, takes a long time the MDS will never stop until it breaks out of this loop due to another condition?

Reply:
Right, that's my concern and is one of the reason I put forward a suggestion for driving this through a dedicated thread.